h2o-genmodel/src/main/java/hex/genmodel/tools/PrintMojo.java

```diff
@@ -189,26 +208,51 @@ private void validateArgs() {
 
   private void run() throws Exception {
     validateArgs();
-    PrintStream os;
     if (outputFileName != null) {
-      os = new PrintStream(new FileOutputStream(new File(outputFileName)));
+      handlePrintingToOutputFile();
+    } else if (outputPngName != null) {
+      handlePrintingToTmpFile();
+    } else {
+      handlePrintingToConsole();
     }
-    else {
-      os = System.out;
+  }
+
+  private void handlePrintingToOutputFile() throws IOException {
+    Path dotSourceFilePath = Paths.get(outputFileName);
+    try (PrintStream os = new PrintStream(new FileOutputStream(dotSourceFilePath.toFile()))) {
+      generateDotFromSource(os, dotSourceFilePath);
     }
+  }
 
-    if(genModel instanceof SharedTreeGraphConverter){
-      SharedTreeGraphConverter treeBackedModel = (SharedTreeGraphConverter) genModel;
-      final SharedTreeGraph g = treeBackedModel.convert(treeToPrint, null);
-      if (printRaw) {
-        g.print();
-      }
-      g.printDot(os, maxLevelsToPrintPerEdge, detail, optionalTitle, pTreeOptions);
+  private void handlePrintingToTmpFile() throws IOException {
+    Path dotSourceFilePath = Files.createTempFile(tmpOutputFileName, "");
+    try (PrintStream os = new PrintStream(new FileOutputStream(dotSourceFilePath.toFile()))) {
+      generateDotFromSource(os, dotSourceFilePath);
     }
-    else {
+  }
+
+  private void handlePrintingToConsole() throws IOException {
+    generateDotFromSource(System.out, null);
+  }
+
+  private void generateDotFromSource(PrintStream os, Path dotSourceFilePath) throws IOException {
+    if (!(genModel instanceof SharedTreeGraphConverter)) {
       System.out.println("ERROR: Unknown MOJO type");
       System.exit(1);
     }
+
+    SharedTreeGraphConverter treeBackedModel = (SharedTreeGraphConverter) genModel;
+    final SharedTreeGraph g = treeBackedModel.convert(treeToPrint, null);
+    if (printRaw) {
+      g.print();
+    }
+    g.printDot(os, maxLevelsToPrintPerEdge, detail, optionalTitle, pTreeOptions);
+    if (outputPngName != null) {
+      try (FileInputStream is = new FileInputStream(dotSourceFilePath.toFile())) {
+        MutableGraph gr = Parser.read(is);
+        Graphviz.fromGraph(gr).render(Format.PNG).toFile(new File(outputPngName));
```

Comment:
This part takes roughly > 8 seconds (no hotspot kicked in) on my computer. I'd guess it's the default DPI settings. I believe this can be tweaked later.

Reply:
Question is: Can we make it faster ? What is the default DPI ?

Reply:
I've investigated this a little bit and the performance problems seems to be observed with the `GraphvizJdkEngine` only. If the default pipeline is used, another one is chosen (probably a binding to the native library) and the time is significantly reduced ( < 1000 ms). Is there anything preventing us from using the default engine pipeline ?

Reply:
Default DPI is 96:
https://github.com/nidi3/graphviz-java/blob/master/graphviz-java/src/main/java/guru/nidi/graphviz/engine/Graphviz.java#withDefaultDpi

I tried to lower it by `gr.graphAttrs().add("dpi",10);` before calling `Graphviz.fromGraph(gr).render(Format.PNG).toFile(new File(outputPngName));` and also tried to fix the width like `Graphviz.fromGraph(gr).width(200).render(Format.PNG).toFile(new File(outputPngName));` but it doesn't seem to have any impact on speed.

Reply:
There is nothing preventing us. I developed it like this in order to demonstrate how non native graphviz version can be used. Default will check for native (GraphvizCmdLineEngine()) and if graphviz is not installed, it will go for the next engine (GraphvizV8Engine()) and next etc... This seems to be a good approach because if native is installed then it is used and if not, then it is used non native (maybe slower, but without need to install new tool). 

The only “problem” is that there will be logs about it from graphviz-java in console (now there wont be logs about which engine was(not) found to be used, because there will be warnings caused by not including log4j depencency and not setting it up together with graphviz-java).

Reply:
The log content, if it makes sense, is not a bad thing. It could even be there to our own advantage. It would be very useful during debugging. Let's pretend there is a situation where this functionality is not working for an imaginary customer. Having the knowledge of which backend has been used and which was skipped could be very useful. Therefore, as I think about it, I'm not really against this being logged.

Reply:
Let's use the defaults then. By default, it is much, much faster then the JDK backend and it uses JDK as fallback on it's own. Win-win situation.

Reply:
Resolved in 3fc46f1d1f10090356a1077c42ee7bcdbd9fcc90

Reply:
I'll try to merge tomorrow - there is a possible relocation issue that needs to be checked. Otherwise we're good.

Reply:
Thank you.